scripts/live-migration-times.d

```diff
@@ -0,0 +1,139 @@
+ *   changes, this script might break.
+ * - We also assume that each phase has a unique name, which is passed as an
+ *   arugment into the migrate_phase_{begin,end} probes. We use the name as
+ *   a key for tracking the phase deltas. If those names change, or phases are
+ *   added or removed, this script will break.
+ */
+
+
+#pragma D option quiet
+#pragma D option defaultargs
+
+dtrace:::BEGIN
+{
+	if ($$1 == "") {
+		printf("ERROR: propolis-server pid required\n");
+		exit(1);
+	}
+
+	printf("tracing live migration protocol times for pid %d...\n", $1);
+	printf("\n");
+}
+
+propolis$1:::migrate_phase_begin
+{
+	this->phase = copyinstr(arg0);
+	start_times[this->phase] = timestamp;
+
+	if ($$2 == "v") {
+		printf("%-25s %-10s timestamp=%d\n", this->phase, "BEGIN",
+		    start_times[this->phase]);
+	}
+}
+
+propolis$1:::migrate_phase_end
+{
+	this->phase = copyinstr(arg0);
+	this->start = start_times[this->phase];
+	this->end = timestamp;
+
+	if (this->start != 0) {
+		delta[this->phase] = this->end - this->start;
+	} else {
+		printf("WARNING: phase \"%s\" could not be measured\n",
+		    this->phase);
+	}
+
+	if ($$2 == "v") {
+		printf("%-25s %-10s timestamp=%d\n", this->phase, "END",
+		    this->end);
+	}
+}
+
+dtrace:::END
+{
+	this->sync = "MigrateSync";
+	this->pause = "Pause";
+	this->rpush = "RamPush";
+	this->dev = "DeviceState";
+	this->rpull = "RamPull";
+	this->fin = "Finish";
+
+	this->d_sync = delta[this->sync];
+	this->d_pause = delta[this->pause];
+	this->d_rpush = delta[this->rpush];
+	this->d_dev = delta[this->dev];
+	this->d_rpull = delta[this->rpull];
+	this->d_fin = delta[this->fin];
+
+	this->total = 0;
+	this->vcpu_pause = 0;
+
+	/* Print header */
+	if ($$1 != "") {
+		printf("\n\n");
+		printf("%-25s %30s\n", "PHASE", "TIME ELAPSED (usec)");
+	}
+
+	/* Print the values of each phase, if they occurred */
+	if (this->d_sync != 0) {
+		printf("%-25s %30d\n", this->sync, this->d_sync / 1000);
+		this->total += this->d_sync;
+	}
+	if (this->d_pause != 0) {
+		printf("%-25s %30d\n", this->pause, this->d_pause / 1000);
+		this->total += this->d_pause;
+		this->vcpu_pause += this->d_pause;
+	}
+	if (this->d_rpush != 0) {
+		printf("%-25s %30d\n", this->rpush, this->d_rpush / 1000);
+		this->total += this->d_rpush;
+		this->vcpu_pause += this->d_rpush;
+	}
+	if (this->d_dev != 0) {
+		printf("%-25s %30d\n", this->dev, this->d_dev / 1000);
+		this->total += this->d_dev;
+		this->vcpu_pause += this->d_dev;
+	}
+	if (this->d_rpull != 0) {
+		printf("%-25s %30d\n", this->rpull, this->d_rpull / 1000);
+		this->total += this->d_rpull;
+		this->vcpu_pause += this->d_rpull;
+	}
+	if (this->d_fin != 0) {
+		printf("%-25s %30d\n", this->fin, this->d_fin / 1000);
+		this->total += this->d_fin;
+		this->vcpu_pause += this->d_fin;
+	}
+	
+	printf("\n");
+
+	/* Print vCPU pause time */
+	if (this->total !=0 && this->d_pause != 0) {
+		printf("%-25s %30d\n", "MAX vCPU PAUSE TIME",
+		    this->vcpu_pause / 1000);
+	}
```

Comment:
This is maybe more of a "file an issue and fix in another PR" sort of thing, but I think this figure should come with a warning label. IIUC it will indicate the total time that elapsed between when the source paused and when the source decided migration was done, but will (necessarily, since the source and destination are different processes) exclude the time between when the destination completed migration and the time it resumed the CPUs. This latter period can be significant if we have to do things like repair Crucible volumes.

Probably the thing to do here is to add some probes to the VM state driver's `start_vm` routine that describe how long it takes to start entities and when we finally get around to starting vCPUs. But it's probably also easier to wait to do that until all the other dust in the state driver has settled.

Reply:
An additional thought: even if we do this, it seems like it might be mildly annoying to have to stitch together traces from multiple hosts to get all the available information about migration performance. Would it be useful to have the source (optionally) transmit performance statistics to the target as part of the protocol so that the target can report everything in one go? (The additional server-state phase in #345 would be a natural place for this, I think.)

Reply:
>  I think this figure should come with a warning label. IIUC it will indicate the total time that elapsed between when the source paused and when the source decided migration was done, but will (necessarily, since the source and destination are different processes) exclude the time between when the destination completed migration and the time it resumed the CPUs.

Yeah, this is a good point, and one I had thought about, both for the vCPU pause time and the script overall, since it's only tracking these specific pieces of the live migration process (though the `RamPush` phase obviously is the bottleneck for the whole process, generally). I added the vCPU bit here because I thought it might at least be nice to have the script do some amount of math for you to get a sense of how long the guest was paused.

For now, I'm happy to remove the vCPU pause message all together and we can file a new issue to add some probes to the state driver, as you suggested. Then we can add that in when we have better instrumentation there. Thoughts on that?

As for stitching together traces from multiple hosts, that wasn't something I was trying to address yet here -- I was more intending for a development tool that could perhaps be run in CI and could assist me with verifying that the adjustments we are making to the guest TSC as a part of the TSC migration work (#337 and related illumos work, [15467)](https://www.illumos.org/issues/15467). I also haven't looked at #345 yet, so I can't speak to that yet. Maybe we could file a new issue for how to stitch together this data for multiple hosts, to track along with the other work laid out in #324?

Reply:
> For now, I'm happy to remove the vCPU pause message all together and we can file a new issue to add some probes to the state driver, as you suggested. 

> Maybe we could file a new issue for how to stitch together this data for multiple hosts, to track along with the other work laid out in https://github.com/oxidecomputer/propolis/issues/324?

Both these approaches sound good to me. I think it's probably also fine to leave the "time spent paused" message in there more or less as-is, or with a comment in the D script describing its limitations. It's up to you; these are (as you mentioned) developer-facing scripts to begin with, so I'm not concerned about e.g. an operator being led astray here.

Reply:
after trying to write up the limitations of the vCPU pause message, I ended up deciding it was more trouble than it was worth to have that information in there -- all it's really doing is subtracting out the `Sync` phase time anyway.